### ISSUE TITLE
Lots of oddities when declaring parameterized classes

### ISSUE THREAD

**Issue by STRML** [facebook/flow#1171]
I'm trying to shim Bluebird.

I'd like to use the following shim:

```js
// @flow
declare module "bluebird" {
  export default class Promise<R>  {
    constructor(callback: (
      resolve: (result: Promise<R> | R) => void,
      reject:  (error: any) => void
    ) => mixed): void;

    then<U>(
      onFulfill?: (value: R) => Promise<U> | U,
      onReject?: (error: any) => Promise<U> | U
    ): Promise<U>;

    catch<U>(
      onReject?: (error: any) => ?Promise<U> | U
    ): Promise<U>;

    static resolve<T>(object?: Promise<T> | T): Promise<T>;
    static reject<T>(error?: any): Promise<T>;
    static all<T, Elem: Promise<T> | T>(promises: Array<Elem>): Promise<Array<T>>;
    static race<T, Elem: Promise<T> | T>(promises: Array<Elem>): Promise<T>;
  }
}
```

This shim is of course incomplete but it mimics the native promise typedef exactly.

This fails with:

```
interfaces/bluebird.js:4
  4:   declare export default class Promise<R>  {
               ^^^^^^ Library parse error:
  4:   declare export default class Promise<R>  {
               ^^^^^^ Unexpected token export
```

Nope, that's not the syntax...

```
interfaces/bluebird.js:4
  4:   declare default export class Promise<R>  {
               ^^^^^^^ Library parse error:
  4:   declare default export class Promise<R>  {
               ^^^^^^^ Unexpected token default
```

Nope... so I guess you can't do this at all.

So I switch to the undocumented `haste` style (some very basic notes on what this is for those of us not working at Facebook would be very welcome):

```js
// @flow
// @providesModule bluebird
declare export default class Promise<R>  { ...
```

This fails with:

```
[No file]:0
inconsistent use of library definitions
  3: declare export default class Promise<R>  {
                                  ^^^^^^^ export default class Promise {}. Type is incompatible with (unclassified use type: SetNamedExportsT). See: interfaces/bluebird.js:3
module
```

Renaming to `Bluebird` (in case we're clobbering Promise) doesn't change the error.

So I ran into @jeffmo on IRC and he suggested:

```js
declare class default<T>{...
```

which returns `unexpected token default`, so that's out.

@jeffmo suggests:

```js
// @flow
// @providesModule bluebird 

declare class Promise<R> {
  constructor(callback: (
    resolve: (result: Promise<R> | R) => void,
    reject:  (error: any) => void
  ) => mixed): void;

  then<U>(
    onFulfill?: (value: R) => Promise<U> | U,
    onReject?: (error: any) => Promise<U> | U
  ): Promise<U>;

  catch<U>(
    onReject?: (error: any) => ?Promise<U> | U
  ): Promise<U>;

  static resolve<T>(object?: Promise<T> | T): Promise<T>;
  static reject<T>(error?: any): Promise<T>;
  static all<T, Elem: Promise<T> | T>(promises: Array<Elem>): Promise<Array<T>>;
  static race<T, Elem: Promise<T> | T>(promises: Array<Elem>): Promise<T>;
}

declare var exports: {default: Class<Promise>};
```

Which sounds like it could work, while hacky, but this does the oddest thing - my console is filled with errors like:

```
160:   assert(key.length <= 16);
       ^^^^^^^^^^^^^^^^^^^^^^^^ function call. Callable signature not found in
160:   assert(key.length <= 16);
       ^^^^^^ module `assert`
```

... which is nowhere near the code I'm writing! It appears I've managed to clobber the hack used to [define assert()](https://github.com/facebook/flow/blob/d2a4515be5ea1ab06d1a087eb75cb1f2cb9df3a8/lib/node.js#L1119).

A last-ditch effort was something like:

```js
declare module "bluebird" {
  declare class Promise<R>  {
    constructor(callback: (
      resolve: (result: Promise<R> | R) => void,
      reject:  (error: any) => void
    ) => mixed): void;

    then<U>(
      onFulfill?: (value: R) => Promise<U> | U,
      onReject?: (error: any) => Promise<U> | U
    ): Promise<U>;

    catch<U>(
      onReject?: (error: any) => ?Promise<U> | U
    ): Promise<U>;

    static resolve<T>(object?: Promise<T> | T): Promise<T>;
    static reject<T>(error?: any): Promise<T>;
    static all<T, Elem: Promise<T> | T>(promises: Array<Elem>): Promise<Array<T>>;
    static race<T, Elem: Promise<T> | T>(promises: Array<Elem>): Promise<T>;
  }
  // Also tried exports: Promise, exports: {default: Promise}, etc..
  declare var exports: {default: Class<Promise>};
}
```

This simply reports no errors, even on obviously wrong code.

So I've hit a wall - is there any way in 0.19 to declare a parameterized class type as a default export?

**Comment by samwgoldman** [comment]
I was able to get this working with the following code:

.flowconfig
```
[libs]
lib
```

lib/bluebird.js
```javascript
declare module bluebird {
  declare class exports<T> {
    constructor(callback: (
      resolve: (result: exports<T> | T) => void,
      reject: (error: any) => void
    ) => void): void;

    then<U>(
      onFulfill?: (value: T) => exports<U> | U,
      onReject?: (error: any) => exports<U> | U
    ): exports<U>;
  }
}
```

test.js
```javascript
// @flow

import Promise from 'bluebird';

var p1: Promise<string> = new Promise((resolve, reject) => {
  resolve("hello");
});

var p2: Promise<number> = p1.then(str => str);
```

Note that it's `declare class exports<T>`, and I had to be careful to refer to the class as `exports` instead of `Promise`—otherwise you're referencing the builtin Promise class, which will error because classes are nominally typed.

Also note that you'll probably run into variance issues with a homegrown Promise declaration. If possible, try to use the builtin Promise support :)

**Comment by STRML** [comment]
@samwgoldman I'm unable to get that to actually typecheck anything - I just receive `No Errors!` in the console. Is that on 0.19 or master?

**Comment by STRML** [comment]
I was able to get it to work for a very short time with `flow check-contents < test.js` in a minimal example, but it seems **incredibly** brittle and I've since managed to break it without any visible code changes. Might even be whitespace. Not sure what is going on.

**Comment by samwgoldman** [comment]
I tested on 0.19.1, which is the latest release. How did you run my example? I tested in a completely clean environment where those are the only 3 files. If you are testing in a larger environment, it's likely that there is something else going on. As a first step, if you aren't doing so already, can you try to get expected behavior in a minimal environment to eliminate that possibility?

**Comment by STRML** [comment]
Yes - I was running only the files you posted above. Something is very unstable with this method.

**Comment by STRML** [comment]
I've been able to get Bluebird to work somewhat well by defining a file `util/bluebird`, which is something like:

```js
import Promise from 'bluebird';

... copy/paste Bluebird.d.ts, with adjustments for Flow

export default Promise;
```

Then in `.flowconfig`:

```
[options]
module.name_mapper='bluebird' -> '<PROJECT_ROOT>/util/bluebird'
```

This actually works pretty well!  The remaining issue seems to be with async/await. Defining this interface:

```js
import Bluebird from 'util/bluebird';

// Fix async/await for Bluebird
declare function $await<T>(p: Bluebird<T> | T): T;
```

produces errors similar to:

```
models/chat.js:102
102:   return Promise.resolve({success: true});
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ async return
102:   return {success: true};
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Promise. This type is incompatible with
 93: Chat.new = async function(accountID: number, message: string): Promise<{success: boolean}> {
                                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ Promise
```

If I adjust the interface:

```
import Promise from 'util/bluebird';

// Fix async/await for Bluebird
declare function $await<T>(p: Promise<T> | T): T;
```

The errors go away, but it appears it's just not typechecking at all.

**Comment by samwgoldman** [comment]
Async/await uses the builtin Promise implementation, which I believe is the correct behavior. (Browsers don't use Bluebird.) Another reason to prefer the builtin Promise support if you can.

At this point, I think the most helpful thing for me would be if you could publish a repo to GH that I can use to try to reproduce what you're seeing.

**Comment by STRML** [comment]
Agreed: putting something together now.

**Comment by STRML** [comment]
@samwgoldman https://github.com/STRML/flow-bluebird

**Comment by STRML** [comment]
By redefining $await, I appear to have made it work, seems to be an issue with union types?

https://github.com/STRML/flow-bluebird/tree/possibleSolution

Note also the `FLOW` notes in the interface, it appears we have to define multiple definitions in the opposite order of typescript, or perhaps definitions after the first are being ignored?

**Comment by STRML** [comment]
Now that Flow v0.20.1 is out, I've tried again to use the original `$await` definition (to see the new error).

See [the branch and output](https://github.com/STRML/flow-bluebird/tree/flow0201Errors);

In short:

```
test.js:6
  6:   let output = await doubleAsync(a);
                          ^^^^^^^^^^^^^^ Promise. This type is incompatible with
703: declare function $await<T>(p: Promise<T> | T): T;
                                   ^^^^^^^^^^^^^^ union: type application of identifier `Promise` | type parameter `T` of await. See: interfaces/await.js:703
```

Yet redefining `$await` to:

```
declare function $await<T>(p: Promise<T>): T;
```

works. Why would a union type be more restrictive than a single parameterized type?

**Comment by avikchaudhuri** [comment]
Merging with #2171 (other issues seem to have been fixed as of 0.28).
